README.md
# font-store

[![NPM Version](https://img.shields.io/npm/v/font-store.svg?style=flat)](https://npmjs.org/package/font-store)
[![Build Status](https://img.shields.io/travis/CrocoDillon/font-store.svg?style=flat)](https://travis-ci.org/CrocoDillon/font-store)

Base64 encodes web fonts to cache in LocalStorage for high performance.

## Quick Start

Install this module globally and convert a web font from Google:

```bash
$ npm install -g font-store
$ font-store "http://fonts.googleapis.com/css?family=Dosis"
```

If all goes right, a JSON file will be created for you named something like `fonts.<MD5>.woff.json` or `fonts.<MD5>.woff2.json`.

## Advanced Usage

To see a list of available options type:

```bash
$ font-store --help
```

Currently the only option is to specify what format you want, either woff or woff2.

## More information

You can read my article [“Non-blocking web fonts using LocalStorage”](http://crocodillon.com/blog/non-blocking-web-fonts-using-localstorage) or check out a [demo on CodePen](http://codepen.io/CrocoDillon/pen/dkcbs/left/?editors=001).


CHANGELOG
current version: 1.1.1
  Minify and filter CSS before converting fonts
  Optimize base64 encoded output when format is WOFF2

tag: 1.1.0
  Add WOFF2 support (using --format option)

tag: 1.0.0


test/unit/utils.js
var expect    = require('chai').expect,
    utils     = require('../../lib/modules/utils'),
    constants = require('../../lib/constants');

var complexCss =
      '@font-face {\n\
        font-family: \'Dosis\';\n\
        font-style: normal;\n\
        font-weight: 400;\n\
        src: local(\'Dosis Regular\'), url(http://fonts.gstatic.com/400.woff2) format(\'woff2\'), url(http://fonts.gstatic.com/400.woff) format(\'woff\');\n\
      }\n\
      @font-face {\n\
        font-family: \'Dosis\';\n\
        font-style: normal;\n\
        font-weight: 600;\n\
        src: local(\'Dosis SemiBold\'), url(http://fonts.gstatic.com/600.woff2) format(\'woff2\'), url(http://fonts.gstatic.com/600.woff) format(\'woff\');\n\
      }\n',
    minifiedCssWoff = '@font-face{font-family:\'Dosis\';font-weight:400;src:url(http://fonts.gstatic.com/400.woff) format(\'woff\')}@font-face{font-family:\'Dosis\';font-weight:600;src:url(http://fonts.gstatic.com/600.woff) format(\'woff\')}',
    minifiedCssWoff2 = '@font-face{font-family:\'Dosis\';font-weight:400;src:url(http://fonts.gstatic.com/400.woff2) format(\'woff2\')}@font-face{font-family:\'Dosis\';font-weight:600;src:url(http://fonts.gstatic.com/600.woff2) format(\'woff2\')}',
    url = 'http://google.com',
    cssStringNoQuotes     = '@font-face {font-family: \'Dosis\';src: local(\'Dosis Regular\'), local(\'Dosis-Regular\'), url('   + url +   ')',
    cssStringSingleQuotes = '@font-face {font-family: \'Dosis\';src: local(\'Dosis Regular\'), local(\'Dosis-Regular\'), url(\'' + url + '\')',
    cssStringDoubleQuotes = '@font-face {font-family: \'Dosis\';src: local(\'Dosis Regular\'), local(\'Dosis-Regular\'), url("'  + url +  '")';

describe('utils', function() {
  describe('minifyCss', function() {
    it('should minify a css string keeping only woff', function () {
      var minifiedCss = utils.minifyCss(complexCss, 'woff');
      expect(minifiedCss).to.equal(minifiedCssWoff);
    });
    it('should minify a css string keeping only woff2', function () {
      var minifiedCss = utils.minifyCss(complexCss, 'woff2');
      expect(minifiedCss).to.equal(minifiedCssWoff2);
    });
  });

  describe('extractRegExp', function() {
    it('should extract urls found in a css string', function () {
      var extractedUrls = utils.extractRegExp(cssStringNoQuotes, constants.urlFormat);
      expect(extractedUrls).to.deep.equal([url]);
    });
    it('should also work with single quotes', function () {
      var extractedUrls = utils.extractRegExp(cssStringSingleQuotes, constants.urlFormat);
      expect(extractedUrls).to.deep.equal([url]);
    });
    it('should also work with double quotes', function () {
      var extractedUrls = utils.extractRegExp(cssStringDoubleQuotes, constants.urlFormat);
      expect(extractedUrls).to.deep.equal([url]);
    });
  });

  describe('replaceWithUrlEncodedContent', function() {
    it('should replace urls with base64 encoded content', function () {
      var result = utils.replaceWithUrlEncodedContent(cssStringNoQuotes, url, 'BASE64DATA', 'CONTENTTYPE'),
          indexOfUrl  = result.indexOf(url),
          indexOfData = result.indexOf('BASE64DATA');
      expect(indexOfUrl).to.equal(-1);
      expect(indexOfData).to.not.equal(-1);
    });
  });
});


test/integration/main.js
var expect = require('chai').expect
var fs = require('fs');

var fontStore = require('../../lib/');
var dosisJson = require('../fonts-fixtures/Dosis.json');

var createdFile;

describe('font-store', function () {
  it('should convert the font Dosis to json', function (done) {
    fontStore('http://fonts.googleapis.com/css?family=Dosis', function(fontJson){
      createdFile=fontJson.md5;
      expect(fontJson.md5).to.equal(dosisJson.md5);
      done();
    });
  });

  afterEach(function() {
    if( createdFile ){
      fs.unlink('fonts.'+createdFile+'.woff.json', function (err) {
        if( err ) console.log( err );
      });
    }
    createdFile=null;
  });
});

test/integration/converter.js
var expect = require('chai').expect
var fs = require('fs');

var converter = require('../../lib/converter');
var dosisJson = require('../fonts-fixtures/Dosis.json');

var createdFile;

describe('converter', function() {
  describe('conversion of font Dosis', function() {
    it('should match the md5 sum', function(done) {
      converter.convert('http://fonts.googleapis.com/css?family=Dosis', 'woff', function(err, fontJson) {
        createdFile = fontJson.md5;
        expect(fontJson.md5).to.equal(dosisJson.md5);
        done();
      });
    });
    it('should match the converted font', function(done) {
      converter.convert('http://fonts.googleapis.com/css?family=Dosis', 'woff', function(err, fontJson) {
        createdFile = fontJson.md5;
        expect(fontJson.value).to.equal(dosisJson.value);
        done();
      });
    });
  });

  describe('conversion of invalid font', function() {
    it('should fail', function(done) {
      converter.convert('http://fonts.googleapis.com/css?family=Unexisting font', 'woff', function(err, fontJson) {
        expect(err).not.to.be.null;
        done();
      });
    });
  });

  afterEach(function() {
    if (createdFile) {
      fs.unlink('fonts.' + createdFile + '.woff.json', function(err) {
        if (err && err.code !== 'ENOENT')
          console.log(err);
      });
    }
    createdFile=null;
  });
});


lib/converter.js
var http      = require('http'),
    url       = require('url'),
    crypto    = require('crypto'),
    fs        = require('fs'),
    async     = require('async'),
    utils     = require('./modules/utils'),
    constants = require('./constants');

module.exports = {
  convert: convert
};

function convert(unparsedUrl, format, callback) {
  callback = callback || function() {};

  var parsedUrl = url.parse(unparsedUrl, true),
      options = {
        host: parsedUrl.host,
        port: parsedUrl.port,
        path: parsedUrl.path,
        headers: {
          'User-Agent': constants.userAgent
        }
      },
      css = '';

  http.get(options, function(res) {
    res.setEncoding('utf8');
    res.on('data', function(chunk) {
      css += chunk;
    });
    res.on('end', function() {
      var urls;

      css = utils.minifyCss(css, format);
      urls = utils.extractRegExp(css, constants.urlFormat);

      if (urls.length == 0) {
        callback(new Error('invalid url'));
        return;
      }

      var encodingQueue = urls.map(encodeUrlContent);
      async.parallel(encodingQueue, function(err, results) {
        for (var i = 0; i < results.length; i++) {
          var result = results[i];
          css = utils.replaceWithUrlEncodedContent(css, result.url, result.encodedContent, constants[format].contentType);
        };

        var json = {
          md5: crypto.createHash('md5').update(css).digest('hex'),
          value: css
        };
        callback(null, json);
      })
    })
  });
}

function encodeUrlContent(url) {
  return function(done) {
    http.get(url, function(res) {
      var data = [];
      res.on('data', function(chunk) {
        data.push(chunk);
      });
      res.on('end', function() {
        var encodedContent = Buffer.concat(data).toString('base64');
        done(null, {
          url: url,
          encodedContent: encodedContent,
          contentType: res.headers['content-type']
        });
      });
    });
  }
}


lib/font-store.js
var http   = require('http'),
    url    = require('url'),
    crypto = require('crypto'),
    fs     = require('fs');

function convert(unparsedUrl) {
  var parsedUrl = url.parse(unparsedUrl),
      options = {
        host: parsedUrl.host,
        port: parsedUrl.port,
        path: parsedUrl.path,
        headers: {
          // Need a header to get WOFF back
          'User-Agent': 'Mozilla/5.0 (Windows NT 6.3; WOW64; rv:31.0) Gecko/20100101 Firefox/31.0'
        }
      },
      css = '',
      count = 0,
      base64 = {};

  http.get(options, function(res) {
    res.setEncoding('utf8');
    res.on('data', function(chunk) {
      css += chunk;
    });
    res.on('end', function() {
      css.replace(/url\((.*?)\)/g, function(match, url) {
        count++;
        http.get(url, function(res) {
          var data = [];
          res.on('data', function(chunk) {
            data.push(chunk);
          });
          res.on('end', function() {
            data = Buffer.concat(data);
            base64[url] = data.toString('base64');
            count--;
            if (count === 0) {
              // Done fetching all WOFF files!
              css = css.replace(/url\((.*?)\)/g, function(match, url) {
                return 'url(data:application/x-font-woff;base64,' + base64[url] + ')';
              });
              var json = {
                md5: crypto.createHash('md5').update(css).digest('hex'),
                value: css
              };
              fs.writeFile('fonts.' + json.md5 + '.woff.json', JSON.stringify(json), function(err) {
                if (err) {
                  console.log(err);
                } else {
                  console.log('Done!')
                }
              });
            }
          });
        });
        return match;
      });
    })
  });
}

module.exports = convert;


lib/modules/utils.js
module.exports = {
  minifyCss: minifyCss,
  extractRegExp: extractRegExp,
  replaceWithUrlEncodedContent: replaceWithUrlEncodedContent
};

function minifyCss(cssString, format) {
  var regex = /@font-face\s*\{([^]+?)\}/gi,
      minifiedCss = '',
      fontFace;
  // Find all @font-face blocks
  while ((fontFace = regex.exec(cssString)) !== null) {
    minifiedCss += '@font-face{' +
      fontFace[1]
        // Split to get an array of declarations for this font-face.
        .split(';')
        .map(function(declaration) {
          // Each declaration should be a `property: value` pair.
          declaration  = declaration.split(/:(?!\/\/)/);
          if (declaration.length !== 2) {
            return '';
          }
          var property = declaration[0].trim(),
              value    = declaration[1].trim();
          if (property === 'src') {
            // The `src` property is special, we only need the url where the format
            // matches the requested format and not any other formats or local fonts.
            value = value
              .split(',')
              .filter(function(urlProperty) {
                var match = /format\(['"]?(.*?)['"]?\)/g.exec(urlProperty);
                return match && match[1] === format;
              })[0].trim();
          }
          return [property, value].join(':');
        })
        .filter(function(declaration) {
          // Filters out any empty or default declarations.
          return declaration && declaration !== 'font-style:normal';
        })
        // Join all declarations back together.
        .join(';') +
    '}';
  }
  return minifiedCss;
}

function extractRegExp(cssString, format) {
  var matches = [],
      match;
  while ((match = format.exec(cssString)) !== null) {
    matches.push(match[1]);
  }
  return matches;
}

function replaceWithUrlEncodedContent(css, url, encodedContent, contentType) {
  return css.replace(url, 'data:' + contentType + ';base64,' + encodedContent);
}


lib/index.js
var http      = require('http'),
    url       = require('url'),
    crypto    = require('crypto'),
    fs        = require('fs'),
    utils     = require('./modules/utils'),
    constants = require('./constants'),
    async     = require('async');

function convert(unparsedUrl,callback) {
  callback = callback || function(){};
  var parsedUrl = url.parse(unparsedUrl),
      options = {
        host: parsedUrl.host,
        port: parsedUrl.port,
        path: parsedUrl.path,
        headers: {
          // Need a header to get WOFF back
          'User-Agent': 'Mozilla/5.0 (Windows NT 6.3; WOW64; rv:31.0) Gecko/20100101 Firefox/31.0'
        }
      },
      css = '';

  http.get(options, function(res) {
    res.setEncoding('utf8');
    res.on('data', function(chunk) {
      css += chunk;
    });
    res.on('end', function() {
      var urls = utils.extractRegExp(css, constants.urlFormat);
      var encodingQueue = urls.map(encodeUrlContent);
      async.parallel(encodingQueue,function(err,results){
        for (var i = 0; i < results.length; i++) {
          var result = results[i];
          css = replaceWithUrlEncodedContent(css,result.url,result.encodedContent);
        };

        var json = {
          md5: crypto.createHash('md5').update(css).digest('hex'),
          value: css
        };
        callback(json);
        fs.writeFile('fonts.' + json.md5 + '.woff.json', JSON.stringify(json), function(err) {
          if (err) {
            console.log(err);
          } else {
            console.log('Done!')
          }
        });
      })
    })
  });
}
function encodeUrlContent(url){
  return function(done) {
    http.get(url, function(res) {
      var data = [];
      res.on('data', function(chunk) {
        data.push(chunk);
      });
      res.on('end', function() {
        var encodedContent = Buffer.concat(data).toString('base64');
        done(null,{
          url:url,
          encodedContent:encodedContent
        });
      });
    });
  }
}

function replaceWithUrlEncodedContent(css,url,encodedContent){
  return css.replace(url,'data:application/x-font-woff;base64,' + encodedContent);
}

module.exports = convert;